lxd/auth/driver_tls.go

```diff
@@ -1,65 +1,419 @@
 }
 
-func (a *tls) UserAccess(username string) (*UserAccess, error) {
-	return &UserAccess{Admin: true}, nil
+// UserAccess is not supported.
+func (a *tls) UserAccess(ctx context.Context, username string) (*UserAccess, error) {
+	return nil, ErrNotSupported
 }
 
-// UserIsAdmin checks whether the requestor is a global admin.
-func (a *tls) UserIsAdmin(r *http.Request) bool {
+// UserHasPermission checks whether the requestor has a specific permission on a project.
+func (a *tls) UserHasPermission(ctx context.Context, r *http.Request, object string, entitlement Entitlement) bool {
 	val := r.Context().Value(request.CtxAccess)
 	if val == nil {
 		return false
 	}
 
-	ua := val.(*UserAccess)
-	return ua.Admin
+	ua, ok := val.(*UserAccess)
+	if !ok {
+		return false
+	}
+
+	if ua.Admin {
+		return true
+	}
+
+	permission, projectName, allow := a.relationToPermission(object, entitlement)
+	if allow {
+		return true
+	}
+
+	return shared.StringInSlice(string(permission), ua.Projects[projectName])
 }
 
-// UserHasPermission checks whether the requestor has a specific permission on a project.
-func (a *tls) UserHasPermission(r *http.Request, projectName string, permission string) bool {
+// GetPermissionChecker returns a function that can be used to check whether a user has the required entitlement on an authorization object.
+func (a *tls) GetPermissionChecker(ctx context.Context, r *http.Request, entitlement Entitlement, objectType ObjectType) (func(object string) bool, error) {
+	allowFunc := func(b bool) func(string) bool {
+		return func(string) bool {
+			return b
+		}
+	}
+
 	val := r.Context().Value(request.CtxAccess)
 	if val == nil {
-		return false
+		return allowFunc(false), nil
+	}
+
+	ua, ok := val.(*UserAccess)
+	if !ok {
+		return allowFunc(false), nil
 	}
 
-	ua := val.(*UserAccess)
 	if ua.Admin {
-		return true
+		return allowFunc(true), nil
+	}
+
+	return func(object string) bool {
+		permission, projectName, allow := a.relationToPermission(object, entitlement)
+		if allow {
+			return true
+		}
+
+		return shared.StringInSlice(string(permission), ua.Projects[projectName])
+	}, nil
+}
+
+func (a *tls) relationToPermission(object string, entitlement Entitlement) (Permission, string, bool) {
+	objectTypeStr, objectName, ok := strings.Cut(object, ":")
+	if !ok {
+		return "", "", false
+	}
+
+	// Project names cannot contain an underscore, so we can extract the project name here.
+	var projectName string
+	potentialProjectName, _, ok := strings.Cut(objectName, "_")
+	if ok {
+		projectName = potentialProjectName
+	}
+
+	objectType := ObjectType(objectTypeStr)
+
+	switch objectType {
+	case ObjectTypeServer:
+		switch entitlement {
+		case EntitlementCanEdit:
+			return PermissionAdmin, "", false
+		case EntitlementCanView:
+			return "", "", true
+		case EntitlementCanCreateStoragePools:
+			return PermissionAdmin, "", false
+		case EntitlementCanCreateProjects:
+			return PermissionAdmin, "", false
```

Comment:
Should the permission for creating projects be "admin" or "manage-projects"? There is currently (in main) no access handler defined for POST /1.0/projects which means project creation is currently restricted to admins only.